docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+
+## Structure
+
+A CSDL document must be a valid [GraphQL schema definition language](https://spec.graphql.org/draft/#sec-Type-System) document.
+
+CSDL removes certain SDL features, as described in the [divergence](#divergence-from-graphql-sdl) section. Nevertheless, CSDL consumers **should** be prepared to parse any valid GraphQL SDL. A CSDL document which contains disallowed SDL constructs **should** trigger validation errors after parsing.
+
+CSDL introduces a suite of types and directives. All CSDL-specific types and directives are prefixed with `csdl_` as recommended in [the GraphQL specification](https://spec.graphql.org/draft/#note-ca863).
+
+### Divergence from GraphQL SDL
+Formally, CSDL is a subset of the [GraphQL schema definition language](https://spec.graphql.org/draft/#sec-Type-System).
+
+In particular, CSDL differs from GraphQL SDL in the following ways:
+- **Extensions are forbidden.** You cannot `extend type` or `extend interface` within a CSDL document.
+- **`schema @csdl(version:)` is mandatory.** Conforming CSDL documents must have a `schema` declaration which must be annotated with a `@csdl` directive specifying, at a minimum, the semver for the CSDL spec in use.
+- the CSDL must define certain types as described in the [mandatory types](#mandatory-types) section.
+
+### Mandatory Types
+
+A CSDL document **must** define a `csdl_Graph` enum. Each enum value defines a subgraph. Each value **must** be annotated with a `@csdl_endpoint` directive specifying the endpoint URL for the subgraph.
+
+```graphql
+enum csdl_Graph {
+  ROCKETS @csdl_endpoint(url: "https://rockets.api.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.api.com"),
+}
+```
+
+The `csdl_Graph` enum is used as input to csdl directives which link fields and types to subgraphs.
+
+## Scalars
+
+### `scalar csdl_SelectionSet`
+
+A GraphQL selection set represented with the same syntax as defined in [the GraphQL SDL](https://spec.graphql.org/draft/#sec-Selection-Sets).
+
+Note: Unlike the federation `_FieldSet` scalar, a `csdl_SelectionSet` must parse as a GraphQL selection set, so the selections must be surrounded by braces.
+
+### `scalar csdl_Url`
+
+A [URL](https://www.w3.org/Addressing/URL/url-spec.html).
+
+### `scalar csdl_Version`
+
+A [Semantic versioning 2.0.0](https://semver.org/spec/v2.0.0.html) verison or range.
+
+## Data Model
+TK, draws heavily from federation, a little more general.
+
+### Realized types
+
+A type is *realized* wherever a resolver from a particular *subgraph* returns that type. We use the notation `subgraph::Type` to represent `Type` realized by `subgraph` and `subgraph::Type.field` to represent the resolver for `Type.field` within `subgraph`.
+
+<a class="listing">Query with resolvers and realized types</a>
+```graphql
+query {          # resolver                    | realizes
+                 # ----------------------------|--------------
+  rockets {      # rockets::Query.rockets     -> rockets::Rocket
```

Comment:
This small change would help me to understand things better 🤷 
```suggestion
  rockets {      # rockets::Query.rockets     -> rockets::[Rocket!]!
```

Reply:
I need to flesh this out so it's clearer, but the point here is to conceptually break a type (`Rocket`) into different selection namespaces based on what you can actually select there.

Fields which are on a different type, or which are not known to the `rockets` subgraph, can't be queried directly.

`Query.rockets` runs in the `rockets` subgraph and returns `[Rocket!]!`. Within the selection set, the router can query any fields on `Rocket` which are known to the `rockets` subgraph. Thus, the real type is `rockets::Rocket`. In fact, all type variants—`Rocket`, `[Rocket]`, `Rocket!`, `[Rocket!]!`—have the same realized type, because it's *just* about what you can select.